lib/ThreeElementVectorLoweringPass.cpp

```diff
@@ -0,0 +1,1419 @@
+      }
+
+      auto *EquivalentGV = new GlobalVariable(
+          M, EquivalentTy, GV.isConstant(), GV.getLinkage(),
+          EquivalentInitializer, "",
+          /* insert before: */ &GV, GV.getThreadLocalMode(),
+          GV.getAddressSpace(), GV.isExternallyInitialized());
+      EquivalentGV->takeName(&GV);
+      EquivalentGV->setAlignment(GV.getAlign());
+      EquivalentGV->copyMetadata(&GV, /* offset: */ 0);
+      EquivalentGV->copyAttributesFrom(&GV);
+
+      LLVM_DEBUG(dbgs() << "Mapping global variable:\n\toriginal: " << GV
+                        << "\n\toriginal type: " << *(GV.getValueType())
+                        << "\n\treplacement: " << *EquivalentGV
+                        << "\n\treplacement type: " << *EquivalentTy << "\n");
+
+      GlobalVariableMap.insert({&GV, EquivalentGV});
+    }
+  }
+
+  bool Modified = !GlobalVariableMap.empty();
+  return Modified;
+}
+
+bool ThreeElementVectorLoweringPass::runOnFunction(Function &F) {
+  LLVM_DEBUG(dbgs() << "Processing " << F.getName() << '\n');
+
+  // Skip declarations.
+  if (F.isDeclaration()) {
+    return false;
+  }
+
+  // Lower the function parameters and return type if needed.
+  // It is possible the function was already partially processed when visiting a
+  // call site. If this is the case, a wrapper function has been created for it.
+  // However, its instructions haven't been visited yet.
+  Function *FunctionToVisit = convertUserDefinedFunction(F);
+  if (FunctionToVisit == nullptr) {
+    // The parameters don't rely on long vectors, but maybe some instructions in
+    // the function body do.
+    FunctionToVisit = &F;
+  }
+
+  bool Modified = (FunctionToVisit != &F);
+  for (Instruction &I : instructions(FunctionToVisit)) {
+    // Use the Value overload of visit to ensure cache is used.
+    Modified |= (visit(static_cast<Value *>(&I)) != nullptr);
+  }
+
+  cleanDeadInstructions();
+
+  LLVM_DEBUG(dbgs() << "Final version for " << F.getName() << '\n');
+  LLVM_DEBUG(dbgs() << *FunctionToVisit << '\n');
+
+  return Modified;
+}
+
+Value *ThreeElementVectorLoweringPass::convertBuiltinCall(
+    CallInst &VectorCall, Type *EquivalentReturnTy,
+    ArrayRef<Value *> EquivalentArgs) {
+  Function *VectorFunction = VectorCall.getCalledFunction();
+  assert(VectorFunction);
+
+  IRBuilder<> B(&VectorCall);
+
+  SmallVector<Value *, 16> Args;
+  for (Value *Arg : EquivalentArgs) {
+    if (Arg->getType()->isVectorTy()) {
+      Value *NewArg = UndefValue::get(
+          VectorType::get(Arg->getType()->getScalarType(), 3, false));
+      Value *tmp = B.CreateExtractElement(Arg, (uint64_t)0);
+      NewArg = B.CreateInsertElement(NewArg, tmp, (uint64_t)0);
+      tmp = B.CreateExtractElement(Arg, (uint64_t)1);
+      NewArg = B.CreateInsertElement(NewArg, tmp, (uint64_t)1);
+      tmp = B.CreateExtractElement(Arg, (uint64_t)2);
+      NewArg = B.CreateInsertElement(NewArg, tmp, (uint64_t)2);
+      Args.push_back(NewArg);
```

Comment:
This seems like it should be a shuffle vector.

Reply:
Done